lib/srv/sess.go

```diff
@@ -1240,43 +1255,42 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
 	return nil
 }
 
-// newRecorder creates a new events.StreamWriter to be used as the recorder
+// newRecorder creates a new events.SessionRecorder to be used as the recorder
 // of the passed in session.
-func newRecorder(s *session, ctx *ServerContext) (events.StreamWriter, error) {
+func newRecorder(s *session, ctx *ServerContext) (events.SessionRecorder, error) {
 	// Nodes discard events in cases when proxies are already recording them.
 	if s.registry.Srv.Component() == teleport.ComponentNode &&
 		services.IsRecordAtProxy(ctx.SessionRecordingConfig.GetMode()) {
-		return &events.DiscardStream{}, nil
+		return events.NewDiscardStream(), nil
 	}
 
-	streamer, err := s.newStreamer(ctx)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	rec, err := events.NewAuditWriter(events.AuditWriterConfig{
+	cfg := events.SessionWriterConfig{
 		// Audit stream is using server context, not session context,
 		// to make sure that session is uploaded even after it is closed
-		Context:      ctx.srv.Context(),
-		Streamer:     streamer,
-		SessionID:    s.id,
-		Clock:        s.registry.clock,
-		Namespace:    s.serverMeta.ServerNamespace,
-		ServerID:     s.serverMeta.ServerID,
-		RecordOutput: ctx.SessionRecordingConfig.GetMode() != types.RecordOff,
-		Component:    teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
-		ClusterName:  ctx.ClusterName,
-	})
+		Context:     ctx.srv.Context(),
+		SessionID:   s.id,
+		Clock:       s.registry.clock,
+		Namespace:   s.serverMeta.ServerNamespace,
+		ServerID:    s.serverMeta.ServerID,
+		Component:   teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
+		ClusterName: ctx.ClusterName,
+	}
+	// Return an SessionWriter with a discarding stream so closing it will
+	// work correctly for tests
+	if ctx.IsTestStub {
```

Comment:
I know you didn't write this, but I do wonder if any of your changes would make it easier to get rid of this. Having any sort of `if isTest` logic in production code is a smell.

Reply:
Good point, I'll try and remove it